Add and use `exactly_one` helper

dstandish opened the pull request:
Python's XOR operator `^` is not very readable, only works with True / False, and only works with two values.  E.g. if you need to test "exactly one of a, b, or c", you cannot do  a ^ b ^ c.  I add an "exactly_one" boolean helper to address all of these shortcomings.

ashb reviewed (commented):
Makes sense, but can we not have a new file just for this?

ashb commented on airflow/utils/boolean.py:
```diff
@@ -0,0 +1,31 @@
+            return False
+        elif arg:
+            has_one = True
+    return has_one
```
```suggestion
    return functools.reduce(operator.xor, map(bool, args))
```

😁

dstandish replied:
lol

potiuk replied:
Nice. And ... it can be parallelised ;)

dstandish replied:
hmm so the that's the thing with that XOR, this fails:
```
>>> functools.reduce(operator.xor, map(bool, ['a', 'a', 'a']))
True
```

(same goes for True in place of 'a')

The issue is that `True ^ True ^ True is True`

ashb replied:
Oh yeah fair point - xor trick doesn't work for more than two args.

potiuk replied:
BTW. @dstandish - you do realise that a ^ b ^ c  is (one_or_three() ) :) ?

dstandish replied:
> BTW. @dstandish - you do realise that a ^ b ^ c is (one_or_three() ) :) ?

yeah it's "exactly one OR all three" -- precisely the problem that my very old-fashioned and not-so-new-fangled code addresses :)

dstandish replied:
like in the 3 case you can do `(a ^ b ^ c) and not (a and b and c)`...
which suggests maybe `return functools.reduce(operator.xor, map(bool, args)) and not functools.reduce(operator.and_, map(bool, args))` would work
however, this is not generally true -- indeed it fails when you have 4 elements!

potiuk commented on airflow/utils/helpers.py:
```diff
@@ -244,3 +244,18 @@ def build_airflow_url_with_query(query: Dict[str, Any]) -> str:
+
+
+def exactly_one(*args):
+    """
```
Challenge accepted:

How about

`return sum(map(bool,args)) == 1`

dstandish replied:
we might have a winner

dstandish replied:
ok tests pass you win the prize.

nothing like a little code golf now and  again

potiuk replied:
Though it's less efficient in certain cases :)

dstandish replied:
yeah mine will fail faster but yours is cleaner 🤷

potiuk replied:
But let me be honest - I cheated: https://stackoverflow.com/questions/22798938/how-to-check-if-exactly-one-value-in-an-iterable-resolves-to-true

uranusjr commented on airflow/utils/helpers.py:
```diff
@@ -244,3 +244,18 @@ def build_airflow_url_with_query(query: Dict[str, Any]) -> str:
+            return False
+        elif arg:
+            has_one = True
+    return has_one
```
This can be problematic if the argument is of a type that can accept a falsy value, or if the default is not falsy (e.g. `NOTSET`). This should Can accept an optional `sentinel=None` keyword argument and check for default-ness with `is` instead.

dstandish replied:
My intention is that the user must provide args that are truthy/falsy  

So in that case you would do 

```python
exactly_one(a != NOTSET, b != NOTSET)

```

etc.... what do you think?

dstandish replied:
so the meaning is, "exactly one is truthy", not "exactly one is set"

dstandish replied:
But to explore that thought...

supposing we parameterize the map function like so:

```python
def exactly_one(*args, func=bool):
    """
    Returns True if exactly one of *args is "truthy", and False otherwise.
    :param args:
    :return:
    """
    return sum(map(func, args)) == 1
```

then in your case you could do

```python
exactly_one(a, b, c, func=lambda x: x != NOTSET)
```

But given that you could achieve the same result, with the simpler function in this way:

```python
exactly_one(*map(lambda x: x !=NOTSET, vals))
```

I think maybe keeping the function simpler is better?

uranusjr replied:
That would work. In that case though, I’d make some of the existing changes more explicit e.g. `exactly_one(execution_date is not None, run_id is not None)` (that’s a bit verbose to me though, so parametrising `exactly_one` should still be considered IMO), and document the function in detail so future contributors don’t fall in this pitfall. On second thought, this falsy-none problem may be an unnecessary pitfall in the first place, indicating we should build safeguards in `exactly_one` in the first place.

dstandish replied:
>  That would work. In that case though, I’d make some of the existing changes more explicit e.g. exactly_one(execution_date is not None, run_id is not None)

I went with `exactly_one(execution_date, run_id)` because AFAIK those values obey truthy evaluation.  Are you suggesting that we do `is not None` because we should never do truthy evaluation?  Or because in this particular case there is a reason not to?  When it was XOR there was a reason (namely because XOR doesn't handle truthy), but seems here it should be OK,  and as you pointed out it's less verbose -- wdyt?

> that’s a bit verbose to me though, so parametrising exactly_one should still be considered IMO

I did explore the idea above, and arrived at the feeling that it wasn't worth it.  But if you feel strongly we should parameterize please say so and perhaps propose a manner.  I'm not strongly against parameterizing, it just makes it a little more complicated to understand the behavior and to me it seemed not worth it (since it doesn't really make usage more elegant than using map and a lambda).

> On second thought, this falsy-none problem may be an unnecessary pitfall in the first place, indicating we should build safeguards in exactly_one in the first place.

can you clarify?

uranusjr replied:
> I went with `exactly_one(execution_date, run_id)` because AFAIK those values obey truthy evaluation. Are you suggesting that we do `is not None` because we should never do truthy evaluation? Or because in this particular case there is a reason not to?

Because `run_id` can be an empty string, and we want to show a different error for this situation.

> can you clarify?

Basically the above. The main purpose of this function (well, the xor checks it is replacing) is to check whether arguments are passed, not whether those arguments are falsy. And if you don’t either explicitly use `is not None`, or include stricter evaluation logic in `exactly_one`, this creates a subtle semantic difference between the caller site and the body of `exactly_one` that can potentially trip people up and cause subtle, unintended behavioural change. And this situation is complete avoidable by changing the evaluation logic to not rely on falsiness.

dstandish replied:
> Because run_id can be an empty string, and we want to show a different error for this situation.

OK in that case I updated XCom to use `is not None` to ensure no change in behavior.  And out of an abundance of caution, same for emr add steps.  The others did not need it.

> subtle, unintended behavioural change

Yeah I was mindful of the potential that it would be a meaningful change, and did consider leaving the xcom calls with `is not None` to begin with.  But I wanted to put forward the less verbose approach with the expectation that we'd confirm through the PR process whether it's  kosher (as we've now done :) ).

So, there should now be no breaking changes in this PR, and what remains is to decide on whether we should parameterize or not and if so how.  It's also something we could add later.

ashb commented on airflow/utils/helpers.py:
```diff
@@ -244,3 +244,12 @@ def build_airflow_url_with_query(query: Dict[str, Any]) -> str:
+
+def exactly_one(*args):
+    """
+    Returns True if exactly one of *args is "truthy", and False otherwise.
```
```suggestion
    Returns True if exactly one of *args is "truthy", and False otherwise.
    
```

ashb commented on airflow/providers/amazon/aws/operators/emr_add_steps.py:
```diff
@@ -62,7 +63,7 @@ def __init__(
         if kwargs.get('xcom_push') is not None:
             raise AirflowException("'xcom_push' was deprecated, use 'do_xcom_push' instead")
-        if not (job_flow_id is None) ^ (job_flow_name is None):
+        if not exactly_one(job_flow_id is not None, job_flow_name is not None):
```
We can't use this in providers -- otherwise it means the provider will only work with Airflow 2.3+, and while it's possible we can do that, this change isn't worth imposing that limit..

dstandish replied:
yup true forgot about that thanks

dstandish committed:
Add and use `exactly_one` helper

The XOR operator `^` is not very readable, only works with True / False, and only works with two values.  E.g. if you need to test "exactly one of a, b, or c", you cannot do  a ^ b ^ c.  I add an "exactly_one" boolean helper to address these shortcomings.

dstandish committed:
update test to verify that combinations of true and truthy are ok; raise error if user provides an iterable

dstandish committed:
remove unused import
